misc/userscripts/open_download

```diff
@@ -0,0 +1,105 @@
+#!/bin/bash -e
+# Both standalone script and qutebrowser userscript that opens a rofi menu with
+# all files from the download director and opens the selected file. It works
+# both as a userscript and a standalone script that is called from outside of
+# qutebrowser.
+#
+# Suggested keybinding (for "show downloads"):
+# spawn --userscript ~/.config/qutebrowser/open_download
+#     sd
+#
+# Requirements:
+#  - rofi (in a recent version)
+#  - xdg-open and xdg-mime
+#  - You should configure qutebrowser to download files to a single directory
+#  - It comes in handy if you enable remove-finished-downloads. If you want to
+#    see the recent downloads, just press "sd".
+#
+# Thorsten Wißmann, 2015 (thorsten` on freenode)
+# Any feedback is welcome!
+
+# open a file from the download directory using rofi
+DOWNLOAD_DIR=${DOWNLOAD_DIR:-$HOME/downloads}
+# the name of the rofi command
+ROFI_CMD=${ROFI_CMD:-rofi}
+ROFI_ARGS=${ROFI_ARGS:-}
+
+msg() {
+    local cmd="$1"
+    shift
+    local msg="$*"
+    if [ -z "$QUTE_FIFO" ] ; then
+        echo "$cmd: $msg" >&2
+    else
+        echo "message-$cmd '${msg//\'/\\\'}'" >> "$QUTE_FIFO"
+    fi
+}
+die() {
+    msg error "$*"
+    if [ -n "$QUTE_FIFO" ] ; then
+        # when run as a userscript, the above error message already informs the
+        # user about the failure, not additional "userscript exited with status
+        # 0" needed.
+        exit 0;
+    else
+        exit 1;
+    fi
+}
+
+if ! [ -d "$DOWNLOAD_DIR" ] ; then
+    die "Download directory »$DOWNLOAD_DIR« not found!"
+fi
+if ! $(which "${ROFI_CMD}" > /dev/null ) ; then
+    die "Rofi command »${ROFI_CMD}« not found in PATH!"
+fi
+
+rofi_default_args=(
+        -monitor -2 # place above window
+        -location 6 # aligned at the bottom
+        -width 100  # use full window width
+        -i
+        -no-custom
+        -format i   # make rofi return the index
+        -l 10
+        -p 'Open download:' -dmenu
+    )
+
+crop-first-column() {
+    local maxlength=${1:-40}
+    local expression='s|^\([^\t]\{0,'"$maxlength"'\}\)[^\t]*\t|\1\t|'
+    sed "$expression"
+}
+
+ls-files() {
+    ls -Q --quoting-style escape    -h  -o -1 -A -t "${DOWNLOAD_DIR}" \
+        | grep '^[-]' \
+        | cut -d' ' -f3- \
+        | sed 's,^\(.*[^\]\) \(.*\)$,\2\t\1,' \
+        | sed 's,\\\(.\),\1,g'
+}
+
+mapfile -t entries < <(ls-files)
+
+line=$(printf "%s\n" "${entries[@]}"  \
+       | crop-first-column 55 \
+       | column -s $'\t' -t \
+       | $ROFI_CMD "${rofi_default_args[@]}" $ROFI_ARGS) || true
+if [ -z "$line" ];  then
```

Comment:
When there are no files (e.g because `DOWNLOAD_DIR` doesn't exist), this condition will be triggered and the script will silently do nothing. That case should probably be handled properly.